internal/blueprint/fsnode_customizations.go

```diff
@@ -0,0 +1,468 @@
+		}
+		file.Group = int64(gid)
+	}
+	// try converting to fsnode.File to validate all values
+	_, err := file.ToFsNodeFile()
+	if err != nil {
+		return err
+	}
+
+	*f = file
+	return nil
+}
+
+// ToFsNodeFile converts the FileCustomization to an fsnode.File
+func (f FileCustomization) ToFsNodeFile() (*fsnode.File, error) {
+	var data []byte
+	if f.Data != "" {
+		data = []byte(f.Data)
+	}
+
+	var mode *os.FileMode
+	if f.Mode != "" {
+		err := validateModeString(f.Mode)
+		if err != nil {
+			return nil, err
+		}
+		modeNum, err := strconv.ParseUint(f.Mode, 8, 32)
+		if err != nil {
+			return nil, fmt.Errorf("invalid mode %s: %v", f.Mode, err)
+		}
+		mode = common.ToPtr(os.FileMode(modeNum))
+	}
+
+	return fsnode.NewFile(f.Path, mode, f.User, f.Group, data)
+}
+
+// FileCustomizationsToFsNodeFiles converts a slice of FileCustomization to a slice of *fsnode.File
+func FileCustomizationsToFsNodeFiles(files []FileCustomization) ([]*fsnode.File, error) {
+	if len(files) == 0 {
+		return nil, nil
+	}
+
+	var fsFiles []*fsnode.File
+	var errors []error
+	for _, file := range files {
+		fsFile, err := file.ToFsNodeFile()
+		if err != nil {
+			errors = append(errors, err)
+		}
+		fsFiles = append(fsFiles, fsFile)
+	}
+
+	if len(errors) > 0 {
+		return nil, fmt.Errorf("invalid file customizations: %v", errors)
+	}
+
+	return fsFiles, nil
+}
+
+// ValidateDirFileCustomizations validates the given Directory and File customizations.
+// If the customizations are invalid, an error is returned. Otherwise, nil is returned.
+//
+// It currently ensures that:
+// - No file path is a prefix of another file or directory path
+// - There are no duplicate file or directory paths in the customizations
+func ValidateDirFileCustomizations(dirs []DirectoryCustomization, files []FileCustomization) error {
+	fsNodesMap := make(map[string]interface{}, len(dirs)+len(files))
+	nodesPaths := make([]string, 0, len(dirs)+len(files))
+
+	// First check for duplicate paths
+	duplicatePaths := make([]string, 0)
+	for _, dir := range dirs {
+		if _, ok := fsNodesMap[dir.Path]; ok {
+			duplicatePaths = append(duplicatePaths, dir.Path)
+		}
+		fsNodesMap[dir.Path] = dir
+		nodesPaths = append(nodesPaths, dir.Path)
+	}
+
+	for _, file := range files {
+		if _, ok := fsNodesMap[file.Path]; ok {
+			duplicatePaths = append(duplicatePaths, file.Path)
+		}
+		fsNodesMap[file.Path] = file
+		nodesPaths = append(nodesPaths, file.Path)
+	}
+
+	// There is no point in continuing if there are duplicate paths,
+	// since the fsNodesMap will not be valid.
+	if len(duplicatePaths) > 0 {
+		return fmt.Errorf("duplicate files / directory customization paths: %v", duplicatePaths)
+	}
+
+	invalidFSNodes := make([]string, 0)
+	checkedPaths := make(map[string]bool)
+	// Sort the paths so that we always check the longest paths first. This
+	// ensures that we don't check a parent path before we check the child
+	// path.
+	sort.Sort(sort.Reverse(sort.StringSlice(nodesPaths)))
```

Comment:
I might be overlooking something but does this sort on strlen of each path? If it does it should sort on dirsep-count.

Reply:
You are probably onto something here... 🤔 While the result will be IMO still correct, the optimization does not work as I intended... Will think about this a bit.

Reply:
I think I get what your'e saying but I don't see how it can cause an issue.  We only really care that paths are sorted so child paths are ahead of their own parent paths, which this should do.  A path with fewer components being ahead of a path with more components shouldn't cause issues if the former isn't a parent of the latter, right?

Reply:
Yeah, there is no impact on the result being correct, but the logic is not correct. I'll just soft the slice using a custom comparison function which will use length based on dir separator.